Fix #6629: linkcheck: Handle rate-limiting

francoisfreitag opened the pull request:
### Feature or Bugfix
- Feature

### Purpose
Follow the Retry-After header if present, otherwise use an exponential
back-off.

### Relates
Issue #6629
Issue #7388

francoisfreitag committed:
Fix #6629: linkcheck: Handle rate-limiting

Follow the Retry-After header if present, otherwise use an exponential
back-off.

francoisfreitag commented:
The Windows checks are failing on tests `test_too_many_requests_user_retry_float` ad `test_too_many_requests_retry_exponential_backoff`. I could not think of a way to meaningfully test the retrying other than actually sleep and measure the wait time. I am happy to try your suggestions.

I can increase the margin on Windows to accommodate for the seemingly longer time needed to start a thread, or skip the test on Windows. I would prefer not to remove these tests.

---

I ended up implementing a version with the Requests library, because of https://github.com/sphinx-doc/sphinx/issues/8391#issuecomment-727266266. I prefer to avoid forcing the use of async when not absolutely mandatory. That way, the project can move on without async, until we decide the time is ripe.

I am happy to pursue the async way, I am convinced it will yield better performance and possibly simplify the implementation, but at the expense of:
- authentication scheme support, because the AIOHTTP project does not support digest auth, OAuth, NTLM, etc.
- some backward incompatibilities described in https://github.com/sphinx-doc/sphinx/issues/8391#issuecomment-731566666

tk0miya commented on doc/usage/configuration.rst:
```diff
@@ -2525,6 +2525,64 @@ Options for the linkcheck builder
+
+   .. code-block:: python
+
+      linkcheck_retry_on_rate_limit = {
```
Unfortunately, a confval should not have lambda as its value. It breaks the incremental build feature of sphinx-build because it's not serializable. I think it's enough to provide options to control these behavior. Additionally, it would be better to provide an event for those who'd like to enhance it via code, but I still don't understand that there is who wants to do that.

francoisfreitag replied:
I simplified the implementation to only accept a float value: the maximum timeout. Perhaps that’s enough.

tk0miya commented on sphinx/builders/linkcheck.py:
```diff
@@ -33,10 +37,14 @@
 
 uri_re = re.compile('([a-z]+:)?//')  # matches to foo:// and // (a protocol relative URL)
 
+RateLimit = namedtuple('RateLimit', ('delay', 'next_check'))
```
Could you use `typing.NamedTuple` instead? It allows annotating types of members.

Example:
https://github.com/sphinx-doc/sphinx/blob/ad804647b5c1bbbe3bb4bfd23e23bbf0ee20b2dc/sphinx/config.py#L41-L43

tk0miya commented on sphinx/builders/linkcheck.py:
```diff
@@ -172,23 +181,38 @@ def check_uri() -> Tuple[str, str, int]:
+                elif err.response.status_code == 429:
+                    next_check = self.limit_rate(err.response)
+                    if next_check is not None:
+                        self.wqueue.put((next_check, (uri, docname, lineno)), False)
```
Why do you use nested tuple here? I feel flat tuple is more simple.

francoisfreitag replied:
It better separates the priority number from the data, but a flat tuple works as well. I used a flat tuple in the latest revision.

tk0miya commented on sphinx/builders/linkcheck.py:
```diff
@@ -247,11 +271,89 @@ def check(docname: str) -> Tuple[str, str, int]:
+                self.rqueue.put((uri, docname, lineno, status, info, code))
+            self.wqueue.task_done()
+
+    def limit_rate(self, response):
```
Please add type-hints to here.

tk0miya commented on tests/test_build_linkcheck.py:
```diff
@@ -382,3 +392,198 @@ def test_connect_to_selfsigned_nonexistent_cert_file(app):
+    # Duration which allocated to the thread to do its work.
+    margin = 1.0
+    assert 1.0 <= (timestamps[1] - start).total_seconds() < 1.0 + margin
+    assert 2.0 <= (timestamps[2] - timestamps[1]).total_seconds() < 2.0 + margin
```
How about patching `time.sleep()`? It does not waste real CI time and help us to measure how long the linkcheck module sleeps. I don't know it is available even under threading. But it's fragile to do sleeping in the real world.

francoisfreitag replied:
I agree that sleeping is both fragile and wasting test time. It would like to avoid it, but do not see a way to do it. Mocking `time.time()` leads to very tight coupling with the implementation.

Patching `time.sleep()` cause consumer threads to busy-loop (instead of sleep) until the `next_check` timestamp (the priority of the link). It wastes CPU cycles for no benefits.
We could peek at the priority queue, see the `next_check`s queued, but that’s sensitive to the timing of the test and the consumer threads still need to wait until the `next_check`s.

Another option is to unit test the `limit_rate` function, verify its return value (the next check) for various inputs.
It does not verify the entire system behavior, thus offers less guarantees. It is also more coupled with the implementation. OTOH, it avoids sleeping, some mocking and parsing of the test webserver logs, and allows more extensive testing of `limit_rate`. I pushed a separate commit with what I have in mind.

francoisfreitag commented:
Thanks for the review!

I believe I addressed all feedback and the patch is ready for another look.
The test suite is now passing with the changes suggested in https://github.com/sphinx-doc/sphinx/pull/8467#discussion_r528319354.

tk0miya commented on CHANGES:
```diff
@@ -33,6 +33,8 @@ Features added
   messages when failed to resolve a cross-reference
 * #6914: Emit a detailed warning when failed to resolve a ``:ref:`` reference
+* #6629: The linkcheck builder now handles rate limits. See
+  :confval:`linkcheck_retry_on_rate_limit` for details.
```
Could you add a category to the entry?
```suggestion
* #6629: linkcheck: The linkcheck builder now handles rate limits. See
  :confval:`linkcheck_retry_on_rate_limit` for details.
```

tk0miya commented on doc/usage/configuration.rst:
```diff
@@ -2525,6 +2525,23 @@ Options for the linkcheck builder
+   doubling the wait time between attempts until it succeeds or exceeds the
+   ``linkcheck_rate_limit_timeout``. By default, the timeout is 5 minutes.
+
+   .. _Retry-After: https://tools.ietf.org/html/rfc2616#section-14.37
```
RFC 2616 was obsoleted. RFC 7231 is better now.
https://tools.ietf.org/html/rfc7231#section-7.1.3

tk0miya reviewed (approved):
LGTM with nits.

francoisfreitag commented:
Rebased, applied suggestions (removed duplicate mention of linkcheck in the CHANGES) and squashed the commits.

tk0miya commented:
Merged! Thank you for your great work!
